File: ex97.java
// DFS

class Solution {   
    public static boolean DFS(String s1, String s2, String s3, int p1, int p2) {
        int p3 = p1 + p2;
        
        // If one pointer reaches the end
        if (p1 == s1.length())
            return s2.substring(p2, s2.length()).equals(s3.substring(p3, s3.length()));
        if (p2 == s2.length())
            return s1.substring(p1, s1.length()).equals(s3.substring(p3, s3.length()));
        
        
        boolean result = false;
        // Match next char in s1
        if (s1.charAt(p1) == s3.charAt(p3))
            result = result || DFS(s1, s2, s3, p1 + 1, p2);
        
        // Match next char in s2
        if (s2.charAt(p2) == s3.charAt(p3))
            result = result || DFS(s1, s2, s3, p1, p2 + 1);
        
        return result;
    }
    
    public boolean isInterleave(String s1, String s2, String s3) {
        return DFS(s1, s2, s3, 0, 0);
    }
}

File: code/ex97.cpp
class Solution {
    string s1, s2, s3;
public:
    bool isInterleave(string s1, string s2, string s3) {
        this->s1 = s1;    this->s2 = s2;    this->s3 = s3;
        vector<vector<int>> memo(s1.length() + 1, vector<int>(s2.length() + 1));
        return backtracking(0, 0, 0, memo);
    }

private:
    bool backtracking(int p1, int p2, int p3, vector<vector<int>>& memo) {
        if (p1 == s1.length() && p2 == s2.length() && p3 == s3.length())    return 1;
        if (p1 >= s1.length() && p2 >= s2.length() && p3 < s3.length())    return 0;
        if (p3 >= s3.length() && (p1 < s1.length() || p2 < s2.length()))    return 0;
        
        if (memo[p1][p2] != 0)    return memo[p1][p2] == 1 ? 1 : 0;
        
        bool res = 0;
        if (s1[p1] == s3[p3])
            res |= backtracking(p1 + 1, p2, p3 + 1, memo);
        if (s2[p2] == s3[p3])
            res |= backtracking(p1, p2 + 1, p3 + 1, memo);
        
        memo[p1][p2] = res ? 1 : -1;
        return res;
    }
};